File: algorithms/python/searching/binary_search.py

"""Binary Search Implementation.

Time Complexity: O(log n)
Space Complexity: O(1) iterative, O(log n) recursive

Prerequisite: Sorted array
"""

from typing import List


def binary_search_iterative(arr: List[int], target: int) -> int:
    """Iterative binary search.
    
    Returns: Index of target if found, -1 otherwise.
    Time: O(log n), Space: O(1)
    """
    left, right = 0, len(arr) - 1
    
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    
    return -1


def binary_search_recursive(arr: List[int], target: int, left: int = 0, right: int = None) -> int:
    """Recursive binary search.
    
    Returns: Index of target if found, -1 otherwise.
    Time: O(log n), Space: O(log n)
    """
    if right is None:
        right = len(arr) - 1
    
    if left > right:
        return -1
    
    mid = (left + right) // 2
    if arr[mid] == target:
        return mid
    elif arr[mid] < target:
        return binary_search_recursive(arr, target, mid + 1, right)
    else:
        return binary_search_recursive(arr, target, left, mid - 1)


if __name__ == "__main__":
    arr = [1, 3, 5, 7, 9, 11, 13, 15]
    
    print("Test 1: Iterative")
    print(f"Search 7: {binary_search_iterative(arr, 7)}")
    print(f"Search 6: {binary_search_iterative(arr, 6)}")
    
    print("\nTest 2: Recursive")
    print(f"Search 7: {binary_search_recursive(arr, 7)}")
    print(f"Search 6: {binary_search_recursive(arr, 6)}")

File: algorithms/java/searching/BinarySearch.java
/**
 * Binary Search Implementation.
 * 
 * Search algorithm for sorted arrays using divide-and-conquer.
 * 
 * Time Complexity: O(log n)
 * Space Complexity: O(1) iterative, O(log n) recursive
 * 
 * Prerequisite: Array must be sorted
 */

public class BinarySearch {
    
    /**
     * Iterative binary search.
     * 
     * @param arr sorted array
     * @param target value to find
     * @return index of target if found, -1 otherwise
     * 
     * Time: O(log n)
     * Space: O(1)
     */
    public static int binarySearchIterative(int[] arr, int target) {
        int left = 0, right = arr.length - 1;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;  // Avoid overflow
            
            if (arr[mid] == target) {
                return mid;
            } else if (arr[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        return -1;
    }
    
    /**
     * Recursive binary search.
     * 
     * @param arr sorted array
     * @param target value to find
     * @return index of target if found, -1 otherwise
     * 
     * Time: O(log n)
     * Space: O(log n) for recursion stack
     */
    public static int binarySearchRecursive(int[] arr, int target) {
        return binarySearchRecursiveHelper(arr, target, 0, arr.length - 1);
    }
    
    private static int binarySearchRecursiveHelper(int[] arr, int target, int left, int right) {
        if (left > right) {
            return -1;
        }
        
        int mid = left + (right - left) / 2;
        
        if (arr[mid] == target) {
            return mid;
        } else if (arr[mid] < target) {
            return binarySearchRecursiveHelper(arr, target, mid + 1, right);
        } else {
            return binarySearchRecursiveHelper(arr, target, left, mid - 1);
        }
    }
    
    /**
     * Demo and testing.
     */
    public static void main(String[] args) {
        int[] arr = {1, 3, 5, 7, 9, 11, 13, 15};
        
        System.out.println("Test 1: Iterative");
        System.out.println("Search 7: " + binarySearchIterative(arr, 7));
        System.out.println("Search 6: " + binarySearchIterative(arr, 6));
        
        System.out.println("\nTest 2: Recursive");
        System.out.println("Search 7: " + binarySearchRecursive(arr, 7));
        System.out.println("Search 6: " + binarySearchRecursive(arr, 6));
        
        System.out.println("\nTest 3: Edge cases");
        System.out.println("Search first: " + binarySearchIterative(arr, 1));
        System.out.println("Search last: " + binarySearchIterative(arr, 15));
    }
}
